dmriprep/utils/register.py

```diff
@@ -0,0 +1,170 @@
+"""
+Linear affine registration tools for motion correction.
+"""
+import numpy as np
+import nibabel as nb
+from dipy.align.metrics import CCMetric, EMMetric, SSDMetric
+from dipy.align.imaffine import (
+    transform_centers_of_mass,
+    AffineMap,
+    MutualInformationMetric,
+    AffineRegistration,
+)
+from dipy.align.transforms import (
+    TranslationTransform3D,
+    RigidTransform3D,
+    AffineTransform3D,
+)
+from nipype.utils.filemanip import fname_presuffix
+
+syn_metric_dict = {"CC": CCMetric, "EM": EMMetric, "SSD": SSDMetric}
+
+__all__ = [
+    "c_of_mass",
+    "translation",
+    "rigid",
+    "affine",
+    "affine_registration",
+]
+
+
+def apply_affine(moving, static, transform_affine, invert=False):
+    """Apply an affine to transform an image from one space to another.
+
+    Parameters
+    ----------
+    moving : array
+       The image to be resampled
+
+    static : array
+
+    Returns
+    -------
+    warped_img : the moving array warped into the static array's space.
+
+    """
+    affine_map = AffineMap(
+        transform_affine, static.shape, static.affine, moving.shape, moving.affine
+    )
+    if invert is True:
+        warped_arr = affine_map.transform_inverse(np.asarray(moving.dataobj))
+    else:
+        warped_arr = affine_map.transform(np.asarray(moving.dataobj))
+
+    return nb.Nifti1Image(warped_arr, static.affine)
```

Comment:
@arokem - this is exactly one of the responsibilities of [poldracklab/nitransforms](https://github.com/poldracklab/nitransforms) which will finally be merged into nibabel when ready.

Would you like me to go through dipy and send a PR to use it (when it is closer to be merged into nibabel)?

(please check the code coverage of nitransforms... that was one of the priorities for the project).

Reply:
Yes - nitransforms looks solid. Do I understand correctly that it ultimately uses scipy.ndimage under the hood to do the resampling? Looks like it always does spline interpolation. What happens if you want nearest neighbors? Maybe nitransforms could use dipy.align.imaffine?

Reply:
Yup, it uses scipy.ndimage. You can pass in `order=0` for nn interpolation :+1: .

The idea for nitransforms is to be as lightweight as possible, as it aims to be merged in nibabel (using dipy.align.imaffine would create a circular dependency, I believe).

I'll have a look into dipy.align.imaffine and potentially ask for your permission to upstream code if that makes sense.

Reply:
Yup. Upstreaming code from dipy.align would make sense. There are some gems in there :-)

Reply:
BTW, at some point I will stop depending on scipy and implement interpolation on a single, lightweight package. Reason: allowing for 4d tensor-bspline interpolation (which might not make sense for diffusion, but will be useful in fmri).

On the diffusion side of things, such package would also include polar interpolation for gradients (and that is relevant to dmriprep). Surely dipy has something for this.

Reply:
> On the diffusion side of things, such package would also include polar interpolation for gradients (and that is relevant to dmriprep). Surely dipy has something for this.

I'm not sure it does... but it would be great to have.

Reply:
To clarify @oesteban -- should we go ahead and drop this apply_affine function in favor of  the nitransforms [equivalent](https://github.com/poldracklab/nitransforms/blob/master/nitransforms/linear.py#L356-L446) or should we wait to do that until the interpolation point if that has already been upstreamed?